### ISSUE TITLE
#outboundproxy (HIGH control need)

### ISSUE THREAD

**Issue by gdestuynder** [nubisproject/nubis-proxy#2]
Note: this is an issue to track the work being done or that has already been done as I found no matching entry.

## From the risk record
Introduce http/https proxy allowing specific outbound patterns.

The steps to implement this security control are:

    Create an http/https proxy server (for example squid)
    Grant that proxy server outbound access to the internet
    For a given product or service, task the product developers with enumerating the outbound sites their product initiates connections to
    Render this site list or site list pattern into rules for the proxy server
    Configure the product or service to use the http/https proxy server instead of initiating connections itself
    Change the security group associated with the product or service from one which allows outbound internet access to one which allows outbound connections only to the proxy server.

This process of creating product specific rules and cutting products over to using the proxy service can be done in a decoupled fashion from other products. Each product team can create their own rules (as they are in the best position to know what outbound connections the product makes).

For products which have not yet enumerated the sites they connect to, they can be pointed to the proxy server with a blanket "allow all" rule while they develop their rules. This doesn't help in preventing data exfiltration but it is useful both in building the rules as well as logging potential data exfiltration.

By introducing an http/https proxy server we also gain the ability to log failed outbound connection attempts, thereby detecting compromised machines.

An http/https proxy server also gives us a powerful tool to surgically stop outbound traffic during a security incident, allowing us to potentially keep the product or service up and available while at the same time preventing exfiltration of data on the servers that makeup that product/service.

## Nubis team comments

TL;DR, This will take me approximately one week to implement for all VPCs in all accounts. If this is a top priority (drop everything) I should be able to have it implemented by the end of next week (Sep 24th 2015) as long as I have the decision by COB Thursday (tomorrow).
The Full Story:
As it is configured today[1] in the datacenters, all Bugzilla nodes are
allowed connections through the proxies to anywhere on allowed ports[2].
This translates to, all current outbound connections for Bugzilla nodes
are logged on allowed ports. All other connections are blocked[3].
It is important to note that the only thing that the proxies are doing
for Bugzilla currently is to log outbound connections on the list of
approved ports[2].
To replicate this functionality in Nubis I would need to
  a) Create puppet module for nubis base (no good module on puppetforge,
would basically copy current module from datacenter puppet and modify to
suit)
  b) Build AMI for proxys utilizing a new nubis-proxy GitHub repository
  c) Use existing consul service discovery to expose proxy service to
hosts as a DNS name
  d) Use existing Fluent collectors to ship logs off to the existing S3
log collection buckets
  e) Use consul key/value store to load/update configuration on running
proxies (technically not necessary as current datacenters allows 'all'.
AKA this is additional, potentially unnecessary, plan-b work)
  f) Leverage existing DataDog integration for monitoring (zero effort
required)
This would be much faster than plumbing additional VPNs and attempting
to integrate with datacenter puppetmasters. After a collaboration with
Gozer, we determined that it is non-trivial to attempt to integrate
ephemeral instances with puppet in a reliable way. Additionally the
semi-reliable method we discussed would require more engineering time
than the above proposed method. I can go into great detail on this point if required.
Props to limed for his assistance helping me to understand the current
datacenter proxy deployment.
1. http_access allow src_(Bugzilla) dst_(Bugzilla)
   src (Bugzilla CIDRs) (ie: 10.8.70.0/24, 10.8.82.0/24)
   dst all
2. acl safe_ports port 21 80 443 873 8000 9418
   acl ssl_ports port 443 873 9418
3. http_access deny all (last line in config)

**Comment by gdestuynder** [comment]
I'm currently seeing the squid logs via the SQS connector - as ec2.forward.squid.* on nubis_events_non_prod

**Comment by gdestuynder** [comment]
The squid config itself is at https://github.com/nubisproject/nubis-proxy/blob/master/nubis/puppet/02_squid.pp

**Comment by gdestuynder** [comment]
Current ACLs allow everything except:

Private IPv4 address space + IPv6 loopback: https://github.com/nubisproject/nubis-proxy/blob/master/nubis/puppet/files/deny_subnets.txt

**Comment by gdestuynder** [comment]
Via https://github.com/nubisproject/nubis-proxy/blob/master/nubis/Puppetfile Nubis makes https://github.com/tinnightcap/puppet-squid3 (not under the nubis project github) authoritative for configuration/module running as root/puppet when building the image, I would recommend we fork it back to nubis or point to a specific commit/tag if possible

**Comment by gdestuynder** [comment]
What I think may still be needed:
- enforcement of usage of proxy to the application (@tinnightcap mentioned this should be done in the nubis-nat project so: https://github.com/nubisproject/nubis-ha-nat/issues/2)
- process to add new ACLs, MOC? (@richardweiss)

**Comment by tinnightcap** [comment]
This is actually being tracked in another issue here:
https://github.com/nubisproject/nubis-meta/issues/30

The enforcement will not happen before go-live of Bugzilla. This needs to be accounted for in the risk assessment and this issue revisited and the risk reduced commensurately.

The process for adding rules will not be handled by the MOC. This is an application level configuration that will be handled by the application owner through the source control repository and then deployed through the ci system. This has not been built, and will not be built prior to go-live. There is an interim procedure for this, which is ping myself or gozer or limed.

To be clear, neither of these outstanding tasks are going to be complete prior to go-live. Additionally I will be closing the other issue as the necessary work has been completed as far as the point release is concerned.

**Comment by tinnightcap** [comment]
Additionally the process issue has been opened here:
https://github.com/nubisproject/nubis-proxy/issues/3

Also FYI, the other issue was filed against meta as this repository did not exist when that issue was opened. You can fine a number of security issues filed against meta where the correct repositories do not yet exist or where it makes no logical sense to file them anywhere else. Please check meta prior to opening new security issues for existing risks. I believe we have issues filed for all open security risks. I will be double checking tomorrow at any rate and will file any that are missing due to the ever changing nature of the security assessment.

I would like to suggest we close this issue now, given that the remaining work is being tracked in other issues which will be on the roadmap, given we have time to plan the roadmap next week. If anyone feels this issue should be reopened, please reach out to me on IRC.

**Comment by gdestuynder** [comment]
@tinnightcap I have revisited the likelihood of the risk related to missing controls for the proxy and lowered them accordingly, which results in MEDIUM risk. This also aligns to your plans of fixing this after go-live which is good news :-)

Also, will look at nubis-meta next time, thanks for the tip.

**Comment by tinnightcap** [comment]
@gdestuynder awesome, thanks.
